### PR TITLE
If apply fails, attempt to run destroy

### LINKED ISSUE #59: If apply fails, some infrastructure may remain
If there is a failure during the apply process, some resources may remain in a created state.

This causes problems for repeated tests, as then subsequent applies will fail because of the previously created resources.

It would be better to destroy resources created by the apply step even if it fails.

### PR DESCRIPTION
**Pull request by baolsen**
closes #59

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -12,7 +12,7 @@ classifiers=[
 readme = "README.md"
```
Comment (baolsen) [inline pyproject.toml:13]: Dependencies update is required for the current black hook to work.

File: pytest_terraform/tf.py [hunk 2]
```diff
@@ -72,8 +72,15 @@ def apply(self, plan=True):
...
-        return TerraformState.from_file(self.state_path, self)
+        try:
+            self._run_cmd(apply_args)
+            return TerraformState.from_file(self.state_path, self)
+        except subprocess.CalledProcessError as e:
+            try:
+                # Try to destroy partially applied resources
+                self.destroy()
+            finally:
+                raise e from None
```
Comment (baolsen) [inline pytest_terraform/tf.py:84]: Intentionally ignore any errors from Destroy. In the test case, Destroy happens to fail for the same reason as the apply fails. In production scenarios, Destroy will normally work and destroy anything partially created by the apply step.

File: pytest_terraform/tf.py [hunk 3]
```diff
@@ -266,7 +273,8 @@ def parse_state(
 
         for r in data.get("resources", ()):
```
Comment (baolsen) [inline pytest_terraform/tf.py:274]: If resources are not correctly applied, the list of instances may be empty.

File: tests/test_terraform.py [hunk 4]
```diff
@@ -174,6 +177,32 @@ def test_tf_runner(testdir, tmpdir):
...
+    content     = "foo!"
+    filename = "${path.module}/foo.bar"
+}
+data "local_file" "foo" {
+    filename = "${path.module}/foo_doesnt_exist.bar"
+    depends_on = [
+        local_file.foo
+    ]
+}
+"""
```
Comment (baolsen) [inline tests/test_terraform.py:196]: Simulate a partial apply:
1 TF apply creates some resources (the local file in this case)
2 TF apply fail on some unrelated resource 

Unfortunately `local_file` resource wont fail if the local file exists already, it just overwrites. So I couldn't use it for the second step.
Instead, we force a failure using a data element to read a non existent file. The side effect of this is the destroy step also fails. But I couldn't find another way to simulate a scenario where a partial apply happens.

**Comment by baolsen** [comment]
@kapilt ready for review

File: tests/test_terraform.py (hunk 4 above)
Comment (baolsen) [reply tests/test_terraform.py:196]: Ref https://github.com/hashicorp/terraform-provider-local/issues/150

**Comment by kapilt** [comment]
thanks for the pull request and detailed commentary.

**Comment by baolsen** [comment]
@kapilt CI done

**Review by kapilt — APPROVED** [review]
thanks, this looks great.

**Pull request merged**